### PR TITLE
Fixed #24782 -- Added assertFormValid and assertFormNotValid methods.

### PR DESCRIPTION
**Pull request by jvzammit**
Adds two functions under `django.test.testcases.SimpleTestCase` as instructed in this #4645 that has so far been closed:
* `assertFormValid`
* `assertFormNotValid`

The two functions are inspired by [this comment](https://github.com/django/django/pull/4645#issuecomment-107254019) by @mjtamlyn .

Unit tests have also been added for each function covering the possible use cases.

Since this is my first PR to an open-source project *ever* I'm appreciative of feedback you will (should) have.

P.S. This PR has started off by applying user `delgiudices` work (1st commit).
Will squash commits once review feedback is applied.

### CODE REVIEW

File: django/test/testcases.py [hunk 1]
```diff
@@ -784,6 +784,57 @@ def assertXMLNotEqual(self, xml1, xml2, msg=None):
...
+
+    def assertResponseFormValid(response, expected_url, context_name='form',
+                                msg=None, **kwargs):
+        raise NotImplementedError
+
+    def assertResponseFormInvalid(response, expected_errors,
+                                  allow_other_errors=True, context_name='form',
+                                  status_code=200, msg=None):
+        raise NotImplementedError
+
```
Comment (charettes) [inline django/test/testcases.py]: I guess these methods could be part of another commit?

File: tests/test_utils/tests.py [hunk 2]
```diff
@@ -812,24 +812,127 @@ class AssertFieldOutputTests(SimpleTestCase):
...
             }
         self.assertFieldOutput(MyCustomField, {}, {}, empty_value=None)
 
 
+class AssertFormValidTests(SimpleTestCase):
+    class TestForm(forms.Form):
+        name = forms.CharField(label='Name', max_length=4)
+        age = forms.IntegerField(label='Age')
+
+    form_cls = TestForm
```
Comment (charettes) [inline tests/test_utils/tests.py]: No need to define another attribute. The form class should be accessible as `self.TestForm`.

File: tests/test_utils/tests.py [hunk 3]
```diff
@@ -812,24 +812,127 @@ class AssertFieldOutputTests(SimpleTestCase):
...
+        })
+
+        with self.assertRaises(AssertionError) as cm:
+            self.assertFormNotValid(
+                form,
+                expected_errors=[('age', 'required'), ],
+                allow_other_errors=False)
+
+        self.assertEqual(
+            cm.exception.message, u"""FormNotValid failed. Error(s):
```
Comment (charettes) [inline tests/test_utils/tests.py]: No need for the `u` prefix, we're already importing `unicode_literals`.

File: tests/test_utils/tests.py [hunk 4]
```diff
@@ -812,24 +812,127 @@ class AssertFieldOutputTests(SimpleTestCase):
...
+            'name': 'John',
+            'age': None,
+        })
+
+        with self.assertRaises(AssertionError) as cm:
+            self.assertFormValid(form)
+
+        self.assertEqual(
+            cm.exception.message,
+            u'''Form validation failed. Error(s):
```
Comment (charettes) [inline tests/test_utils/tests.py]: Ditto.

File: tests/test_utils/tests.py [hunk 5]
```diff
@@ -812,24 +812,127 @@ class AssertFieldOutputTests(SimpleTestCase):
...
+            'name': 'John',
+            'age': 7,
+        })
+
+        with self.assertRaises(AssertionError) as cm:
+            self.assertFormNotValid(form)
+
+        self.assertEqual(
+            cm.exception.message,
+            u'FormNotValid failed. Error(s):\n- Form is valid\n')
```
Comment (charettes) [inline tests/test_utils/tests.py]: Ditto.

**Comment by jvzammit** [comment]
@charettes Review feedback applied, thanks for the feedback! I'll squash commits once I get your "go ahead".

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/test_utils/tests.py [hunk 6]
```diff
@@ -4,13 +4,13 @@
 import unittest
 import warnings
 
+from django import forms
```
Comment (timgraham) [inline tests/test_utils/tests.py]: Please keep cleanups like this in a separate commit or pull request to avoid blaming those changes to a commit which contains new features.

File: tests/test_utils/tests.py [hunk 7]
```diff
@@ -812,24 +812,123 @@ class AssertFieldOutputTests(SimpleTestCase):
...
+        })
+        self.assertFormValid(form)
+
+    def test_error_message_with_invalid_form(self):
+        form = self.TestForm(data={
+            'name': 'John',
+            'age': None,
+        })
+
+        with self.assertRaises(AssertionError) as cm:
```
Comment (timgraham) [inline tests/test_utils/tests.py]: use `self.assertRaisesMessage()`

File: docs/topics/testing/tools.txt [hunk 8]
```diff
@@ -1381,6 +1381,24 @@ your test suite.
     ``errors`` is an error string, or a list of error strings, that are
     expected as a result of form validation.
 
+.. method:: SimpleTestCase.assertFormValid(form, msg=None)
+
+    Asserts that a form is valid. if form is not valid. Displays clean output with
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Please use periods after all sentences and add `.. versionadded:: 1.10` annotations as described in the [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). Also a mention in the 1.10 release notes is needed. Be sure to wrap all documentation at 79 characters.

**Comment by timgraham** [comment]
Please uncheck "Patch needs improvement" on the ticket when you update the patch. It's fine to squash commits and use a commit message that matches the pull request title. Thanks!

File: tests/test_utils/tests.py (hunk 6 above)
Comment (jvzammit) [inline tests/test_utils/tests.py]: @timgraham Thank you for reviewing this. I don't understand what is meant "Please keep cleanups like this in a separate commit or pull request". I know it affects previous code but the alternative would have been to import `Form`, `CharField` and `IntegerField` from `django.forms` individually. Can you kindly instruct and what should be done exactly? Please bear with me as this is my first-ever PR.

Comment (timgraham) [inline tests/test_utils/tests.py]: Create a separate pull request or commit in this pull request that precedes the changes for the ticket which makes the import change.

**Comment by jvzammit** [comment]
@timgraham Ready to commit. However, what should I do about the Py3 failures? Thanks.

**Comment by timgraham** [comment]
Use `self.assertRaisesMessage()`

File: tests/test_utils/tests.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
 from django.core.files.storage import default_storage
 from django.core.urlresolvers import NoReverseMatch, reverse
 from django.db import connection, router
-from django.forms import EmailField, IntegerField
+from django.forms import CharField, EmailField, Form, IntegerField
```
Comment (jvzammit) [inline tests/test_utils/tests.py]: @timgraham I just added more individual imports here rather than causing (and requiring) a "refactoring" PR or commit as you had suggested. That doesn't cause touching of previously-existing code. Wrt all other changes, all improvements you suggested were applied. Thanks again.

**Comment by jvzammit** [comment]
Btw @timgraham I'm trying to unset "Patch needs improvement" from the [ticket](https://code.djangoproject.com/ticket/24782) but I think I've forgot how to edit tickets on Trac. Sorry about that.

File: docs/topics/testing/tools.txt [hunk 10]
```diff
@@ -1544,6 +1544,26 @@ your test suite.
...
+
+    .. versionadded:: 1.10
+
+   Asserts that a form is not valid and that errors are ``expected_errors``.
+
+   ``form`` is the form to be validated.
+
+   ``expected_errors`` is a list of (field, code) errors that are expected on the form.
+
+   ``allow_other_errors`` is a boolean that when set (by default) allows other errors not within ``expected_errors`` to be raised by the form.
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Docs should be wrapped at 79 characters.

**Comment by timgraham** [comment]
Please add a mention in the 1.10 release notes too.
About editing the ticket, you need to be logged in to Trac to edit -- could that be issue?

**Comment by jvzammit** [comment]
@timgraham Added mention in release docs, and wrapped docs lines (except for function signature, all current function signatures in docs appear to go over that 79-chars line wrapping rule). Rebased over the latest Django and pushed again.

Also unchecked "patch needs improvement" in Trac once I logged in.

I still can't figure out why the method reference in release notes is failing. I.e. the method references should be found. Any help with that?

Thanks again! Sorry this is taking so long.

**Comment by jvzammit** [comment]
Rebased on latest master to fix conflict, pushed again. Failed again at docs, asked this [question on StackOverflow](http://stackoverflow.com/questions/35019566/sphinx-reference-target-not-found-warning) someone suggests a resolution to this. CC' @timgraham

File: docs/releases/1.10.txt [hunk 11]
```diff
@@ -301,7 +301,7 @@ Templates
 Tests
 ~~~~~
 
-* ...
+* Added methods :meth:`~django.test.testcases.SimpleTestCase.assertFormValid` and :meth:`~django.test.testcases.SimpleTestCase.assertFormNotValid`.
```
Comment (timgraham) [inline docs/releases/1.10.txt]: omit ".testcases" -- the documentation references use the convenience import path to `SimpleTestCase`.

Comment (jvzammit) [inline docs/releases/1.10.txt]: Thanks @timgraham that eliminates the warning locally. Committed the change, rebased to squash last commit and pushed again. Thanks!

**Comment by jvzammit** [comment]
@timgraham Is this ready to merge (finally)?

**Comment by timgraham** [comment]
I'll give it another review when I have a chance. Can you try to rebase your branch so that the "Merge" commits don't appear here?

File: docs/releases/1.10.txt [hunk 12]
```diff
@@ -301,7 +301,7 @@ Templates
 Tests
 ~~~~~
 
-* ...
+* Added methods :meth:`~django.test.SimpleTestCase.assertFormValid` and :meth:`~django.test.SimpleTestCase.assertFormNotValid`.
```
Comment (timgraham) [inline docs/releases/1.10.txt]: docs should be wrapped at 79 characters

**Comment by jvzammit** [comment]
@timgraham Closing this PR in favour of #6046 .
